### ISSUE TITLE
ParseAndCheckFileInProject may return stale results

### ISSUE THREAD

**Issue by auduchinok** [dotnet/fsharp#11291]
Consider the following repro steps for project with two files:

* Parse and check File2.fs (File1.fs is checked as a dependency, results for File1 and File2 get cached)
* Change File1.fs and parse and check it
* Parse and check File2.fs again, it'll have the same stamp (since its source hasn't changed), so it will use the old results

The following code uses the incremental builder cache that is seemingly never updated by the background compiler after a file check:
https://github.com/dotnet/fsharp/blob/8bfb0c7a4e83c9604ddc7c1ed39d3bdd6d70aa03/src/fsharp/service/service.fs#L480

After checking a file only background compiler caches are updated, not the incremental builder ones:
https://github.com/dotnet/fsharp/blob/8bfb0c7a4e83c9604ddc7c1ed39d3bdd6d70aa03/src/fsharp/service/service.fs#L567

This was discussed in https://github.com/dotnet/fsharp/pull/11228#discussion_r597728199, but FCS 39 has different logic there, and a different fix is needed (unless the PR with the fix is merged soon and everything else is considered stable enough for a new FCS release).

**Comment by auduchinok** [comment]
I've checked commits that I based our FCS build on and the issue appeared at some time between:

* https://github.com/dotnet/fsharp/commit/d21e15e361cc0c6f270828dca07d8aa662a89ebd that doesn't have this issue.
* https://github.com/dotnet/fsharp/commit/8bfb0c7a4e83c9604ddc7c1ed39d3bdd6d70aa03 that has.

**Comment by dsyme** [comment]
The comparison is here: https://github.com/dotnet/fsharp/compare/d21e15e361cc0c6f270828dca07d8aa662a89ebd...8bfb0c7a4e83c9604ddc7c1ed39d3bdd6d70aa03

The relevant diff is I believe this one:

![image](https://user-images.githubusercontent.com/7204669/112040704-70ffa480-8b3d-11eb-85d5-90194982c993.png)

@TIHan My understanding is the removal of the TimeStampCache in this is the cause of the bug?  

In any case  https://github.com/dotnet/fsharp/pull/11228 fixes this.  But we may want to help @auduchinok find a fix too

**Comment by dsyme** [comment]
@auduchinok I think this fix is adequate, it puts the logic of `AreCheckResultsBeforeFileInProjectReady` back to where it was

```fsharp
    member builder.AreCheckResultsBeforeFileInProjectReady filename = 
        let slotOfFile = builder.GetSlotOfFileName filename
        let cache = TimeStampCache defaultTimeStamp
        let ctok = AssumeCompilationThreadWithoutEvidence()
        computeStampedReferencedAssemblies cache ctok
        computeStampedFileNames cache ctok
        match tryGetBeforeSlot slotOfFile with
        | Some _ -> true
        | _ -> false
```

This also basically matches what's in #11228, see https://github.com/dotnet/fsharp/pull/11228/files#diff-4cf589945def81188bed84b5466d6c2c6eca933e91427e54ed059f018fb9cec9R1210

**Comment by dsyme** [comment]
@auduchinok reports this bug is also in Visual Studio 16.9.1.   @TIHan I'm not sure we can do anything about this except wait for 16.10

**Comment by dsyme** [comment]
I pushed the patch to this branch: https://github.com/dotnet/fsharp/tree/patches/fcs-39-bug-10538

@TIHan It might make sense for us to add a test case that explicitly gets fixed by that patch (assuming it's correct, I'm pretty sure it is) and then double check that #11228 also passes that test

**Comment by baronfel** [comment]
I can reproduce this in latest Ionide as well (which is based on FCS 39) with the following very simple test harness:

parent.fsx:
```fsharp
module Dependency =
    let value = 10
```

script.fsx:
```fsharp
#load "./parent.fsx"

open Parent

let result = Dependency.value + 20
```

Follow @auduchinok's test steps, but as the change to 'file 1' change `value` in `parent.fsx` to, say, a boolean value.  You expect `result`'s expression to be in error now, but it's not.

**Comment by TIHan** [comment]
> My understanding is the removal of the call to IncrementalBuilder.IsReady plus the use of a TimeStampCache is the cause of the bug.

Yes, that is right. 

https://github.com/dotnet/fsharp/pull/11228 should fix this since it puts the computing timestamps back in place.

**Comment by dsyme** [comment]
> I can reproduce this in latest Ionide as well (which is based on FCS 39) with the following very simple test harness:

@baronfel Do you have a bespoke fork+build of FCS or do you rely solely on the nuget?   (@auduchinok has a bespoke build in Rider and can thus apply the patch bug fix above). If not  would be good to get the FCS 39.x release updated with the given fix.

**Comment by baronfel** [comment]
@dsyme we rely only on FCS39 at the moment so we'd need a patch release here in order to get a fix.  I suppose we could also look at something like a git-reference using Paket, but we still couldn't use main in that case because our long tail of constituent libraries (like Ionide.ProjInfo, Fantomas, FSharpLint, etc) would be broken due to API changes.

**Comment by cartermp** [comment]
Resolved by https://github.com/dotnet/fsharp/pull/11228

**Comment by auduchinok** [comment]
@cartermp Would it be better to keep it open until a relevant test is added?

**Comment by theprash** [comment]
Hi all. Would you consider taking on the extra work to fix this in 16.9 as well? This seems like a serious bug at least as I described it in #11434. When working on any project I find that I'm very frequently having to do the workarounds of getting my errors from build output and then editing files just to see the errors on them in the editor. So I'm sure lots of people must be affected by it.

I have no idea when 16.10 is scheduled for release but based on previous release dates it could be another couple of months. If you know that it is actually much sooner then that, then I'm sure it's fine to wait. Thanks!

**Comment by deyanp** [comment]
+ don't forget Rider, this is killing me already for a week, I need to have this fixes asap ...

**Comment by cartermp** [comment]
We don't have any plans to service FCS or the VS tools. Although this is a regression, it does not meet the bar for servicing our tools. VS 16.10 will release in about a month or so, +/- some days to account for stability issuess.

**Comment by deyanp** [comment]
@cartermp , I just raised this [Rider Issue](https://youtrack.jetbrains.com/issue/RIDER-61730 ) and it is killing my productivity - do you mean nobody else is affected, and I should wait undefined amount of time till it is really resolved?

**Comment by cartermp** [comment]
Our bar for servicing is very high (e.g., crashes an IDE). That is not the case with this bug.

**Comment by baronfel** [comment]
@dsyme we only use nuget versions of FCS (part of why I bang the 'release FCS more often' drum).

**Comment by isaacabraham** [comment]
@deyanp It is defined - it'll come out with 16.10 which is a few weeks away.

@cartermp FYI - although a couple of people on our team mentioned this to me this week, frankly I hadn't even noticed this. Partly this might be because I tend to jump between Ionide, Rider and VS, all of which have slightly different strengths and weaknesses when it comes to this sort of stuff.

**Comment by deyanp** [comment]
@isaacabraham , I am using Rider ... Assuming 16.10 means Visual Studio, I am wondering when Rider will be "fixed". 

The strange thing is that it seems I am the most affected in my team (of total 4) ... Not sure if it is because I am jumping between files more ... even doubled the java vm memory for Rider, invalidated cache 3 times, but still heavily affected by this issue ...

**Comment by auduchinok** [comment]
> even doubled the java vm memory for Rider, invalidated cache 3 times, but still heavily affected by this issue ...

None of this should have any effect. The easiest workaround is to modify the file showing stale results.

We've applied the fix from https://github.com/dotnet/fsharp/issues/11291#issuecomment-804315854 before the release, but either it doesn't help in some cases or there's some version/packaging issue on the Rider side. We're looking into it.

**Comment by cartermp** [comment]
I used @theprash's repro on the latest dev builds of 16.10 and this **is not** fixed.

@TIHan @dsyme can we look into this?

**Comment by smoothdeveloper** [comment]
What is the shortest work around in visual studio 16.9? I'm also troubleshooting my colleagues.

Thanks!

**Comment by dsyme** [comment]
> @TIHan @dsyme can we look into this?

Yes we can

**Comment by dsyme** [comment]
@TIHan Note this from @theprash 's repro

> The exception is when the later file is the very next file in the project. In that case, the errors and types are updated automatically as expected.


We might have only tested this particular case

**Comment by TIHan** [comment]
I apologize that this is still occurring. I'm trying to pin-point exactly what is going on.

**Comment by vzarytovskii** [comment]
@TIHan, do you know if it was fixed by https://github.com/dotnet/fsharp/pull/11460? @auduchinok mentioned that he seen it happening after https://github.com/dotnet/fsharp/pull/11460 was merged.

**Comment by deyanp** [comment]
One thing is for sure, the problem is still there in Rider, and causes a lot of frustration every day ..

**Comment by cartermp** [comment]
@vzarytovskii No, that doesn't appear to resolve the issue, not wholly at least

**Comment by charlesroddie** [comment]
I am still getting stale results in 16.10:
- build a project with edits
- file in dependent project doens't update until edited

This is with in-memory cross project references disabled and stale data for intellisense features disabled.

**Comment by auduchinok** [comment]
> @TIHan, do you know if it was fixed by #11460? @auduchinok mentioned that he seen it happening after #11460 was merged.

I've just found and fixed another related issue on our side (and backported the fix to 2021.1 branch), and I'm not able to reproduce it anymore. I hope it was the last piece, and the issue is fixed.

**Comment by vzarytovskii** [comment]
@TIHan FYI

**Comment by auduchinok** [comment]
Unfortunately, I still can repro it sometimes with the old repro steps that require typing in a subsequent file to get its results updated.

**Comment by TIHan** [comment]
@auduchinok , what are the exact steps to reproduce this? After we merged the latest fixes, we haven't hit the issue.

**Comment by auduchinok** [comment]
@TIHan I think I've seen it twice, but without a stable repro, unfortunately. I'll add details if I manage to track it down.

**Comment by dsyme** [comment]
I will close this and await a new stable repro, @please open a new issue if you get one
